### PR TITLE
Respect the intent of rules authors when returning atomic patches

### PR DESCRIPTION
**Pull request by mlachkar**
In scalafixFileEvaluation, there is a possibility to apply only some patches.
The current issue, is that we were returning the low level patches, by recursively decomposing even Atomic patches.

If the author rule has created an atomic patch that contains two Add patches, we should not decompose it,
and therefore `previewPatches()` can only be called with "atomic" patches

### CODE REVIEW

File: scalafix-tests/unit/src/main/scala/scalafix/test/FqnRule.scala [hunk 1]
```diff
@@ -50,3 +51,17 @@ class SyntacticRuleV1 extends v1.SyntacticRule("SyntacticRuleV1") {
     Patch.addRight(doc.tree, "\nobject SyntacticRuleV1")
   }
 }
+
+class CommentFileRule1 extends v1.SyntacticRule("CommentFileRule1") {
```
Comment (bjaglin) [inline scalafix-tests/unit/src/main/scala/scalafix/test/FqnRule.scala:55]:
```suggestion
class CommentFileNonAtomic extends v1.SyntacticRule("CommentFileNonAtomic") {
```

File: scalafix-tests/unit/src/main/scala/scalafix/test/FqnRule.scala [hunk 2]
```diff
@@ -50,3 +51,17 @@ class SyntacticRuleV1 extends v1.SyntacticRule("SyntacticRuleV1") {
...
+  override def fix(implicit doc: SyntacticDocument): Patch = {
+    Patch.addLeft(doc.tree, "/*") +
+      Patch.addRight(doc.tree, "*/")
+  }
+}
+
+class CommentFileRule2 extends v1.SyntacticRule("CommentFileRule2") {
```
Comment (bjaglin) [inline scalafix-tests/unit/src/main/scala/scalafix/test/FqnRule.scala:62]:
```suggestion
class CommentFileAtomic extends v1.SyntacticRule("CommentFileAtomic") {
```

File: scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala [hunk 3]
```diff
@@ -214,6 +213,74 @@ class ScalafixArgumentsSuite extends AnyFunSuite with DiffAssertions {
...
+        List(
+          "CommentFileRule1"
+        ).asJava
+      )
+      .withSourceroot(src)
+
+    val fileEvaluation1 = run1.evaluate().getFileEvaluations.head
+    val patches1 = fileEvaluation1.getPatches
+    assert(patches1.length == 2)
+    val obtained1 = fileEvaluation1.previewPatches().get
```
Comment (bjaglin) [inline scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala:228]: no assertion?

File: scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala [hunk 4]
```diff
@@ -214,6 +213,74 @@ class ScalafixArgumentsSuite extends AnyFunSuite with DiffAssertions {
     assert(contentAfterRun == fileEvaluation.previewPatches().get)
   }
 
+  test("ScalafixArguments.evaluate retrieve only atomic patches") {
```
Comment (bjaglin) [inline scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala:216]: I think tests would be more readable (and avoid 1/2 suffixes in variables) if you had more tests with less assertions

File: scalafix-core/src/main/scala/scalafix/v0/Rule.scala [hunk 5]
```diff
@@ -83,7 +83,7 @@ abstract class Rule(ruleName: RuleName) { self =>
     // This overload of apply if purely for convenience
     // Use `applyAndLint` to iterate over Diagnostic without printing to the console
     val result =
-      PatchInternals(patches, ctx, semanticOption)
+      PatchInternals(patches, ctx, semanticOption, suppress = false)
```
Comment (bjaglin) [inline scalafix-core/src/main/scala/scalafix/v0/Rule.scala:86]: are changes in this file related to the other one above?

File: scalafix-core/src/main/scala/scalafix/internal/patch/PatchInternals.scala [hunk 6]
```diff
@@ -191,6 +200,20 @@ object PatchInternals {
     loop(patch)
   }
 
+  // don't decompose Atomic Patch
+  def foreachPatchUnit(patch: Patch)(f: Patch => Unit): Unit = {
```
Comment (bjaglin) [inline scalafix-core/src/main/scala/scalafix/internal/patch/PatchInternals.scala:204]: this is a very generic signature, yet it's used only for accumulation, maybe the 2 new methods could be merged?

**Review by bjaglin — COMMENTED** [review]
Even with `?w=1` the diff is really hard to follow, and I feel like many changes are unnecessary or could be extracted to no-op refactoring commits. Is there any chance you could amend it to try to produce a minimal diff? Thanks!

File: scalafix-core/src/main/scala/scalafix/v0/Rule.scala (hunk 5 above)
Comment (mlachkar) [reply scalafix-core/src/main/scala/scalafix/v0/Rule.scala:86]: no. you are right

File: scalafix-core/src/main/scala/scalafix/internal/patch/PatchInternals.scala (hunk 6 above)
Comment (mlachkar) [reply scalafix-core/src/main/scala/scalafix/internal/patch/PatchInternals.scala:204]: I use it twice

File: scalafix-cli/src/main/scala/scalafix/internal/v1/MainOps.scala [hunk 7]
```diff
@@ -374,7 +374,7 @@ object MainOps {
       ctx: RuleCtx,
       index: Option[v0.SemanticdbIndex]
   ): Option[String] =
-    Try(tokenPatchApply(ctx, index, patches)).toOption
+    Try(patchApply(ctx, index, patches)).toOption
```
Comment (mlachkar) [inline scalafix-cli/src/main/scala/scalafix/internal/v1/MainOps.scala:377]: this is in fact a rename.

**Review by mlachkar — COMMENTED** [review]
yes, I will amend to reduce the diff.

File: scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala (hunk 3 above)
Comment (mlachkar) [reply scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala:228]: I didi an assertion 
`assertNoDiff(obtained1, obtained2)`

File: scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala (hunk 4 above)
Comment (mlachkar) [reply scalafix-tests/unit/src/test/scala/scalafix/tests/interfaces/ScalafixArgumentsSuite.scala:216]: ok

**Comment by mlachkar** [comment]
I removed all my small refactoring (renames, and removing `return` keyword). 
The only change is about not using `underlying` that loop inside an `Atomic Patch`.

**Review by bjaglin — APPROVED** [review]
thanks for taking the time to amend, much clearer :+1:
